posts/Gunnar/2019-01-23-testing-cdi-beans-and-persistence-layer-under-java-se.adoc

```diff
@@ -0,0 +1,456 @@
+so a file _META-INF/services/org.jboss.weld.bootstrap.api.Service_ with the fully-qualified name of our implementation as its contents is required:
+
+[source,java]
+----
+org.hibernate.demos.jpacditesting.support.TestingTransactionServices
+----
+
+And with that, we can now test code like this which makes use of transactional observers:
+
+[source,java]
+----
+@ApplicationScoped
+public class SomeObserver {
+
+    public void observes(@Observes(during=TransactionPhase.AFTER_COMPLETION) String event) {
+        // handle event ...
+    }
+}
+----
+
+We also can use JTA's `@Transactional` annotation to benefit from declarative transaction control:
+
+[source,java]
+----
+@ApplicationScoped
+public class TransactionalGreetingService {
+
+    @Transactional(TxType.REQUIRED)
+    public String greet(String name) {
+        // ...
+    }
+}
+----
+
+When this `greet()` method is invoked, it must be run in a transaction context,
+which either has been started before or will be started if needed.
+Now, if you have used transactional CDI beans before, you might wonder where the associated method interceptor is.
+As it turns out, Narayana comes with CDI support and provides us with everything needed: method interceptors for the different transactional behaviours (`REQUIRED`, `MANDATORY` etc.)
+as well as a portable extension which registers the interceptors with the CDI container.
+
+== Configuring the Weld Initiator
+
+There's one last detail we have ignored so far, and that is how Weld will detect all the beans we need for our test,
+be it actual components under test such as `GreetingService` or testing infrastructure such as the `EntityManagerProducer`.
+The simplest is to let Weld scan the classpath itself and pick up all beans it finds.
+This is enabled by passing a new `Weld` instance to the `WeldInitiator` rule:
+
+[source,java]
+----
+public class CdiJpaTest {
+
+    @Rule
+    public WeldInitiator weld = WeldInitiator.from(new Weld())
+        .activate(RequestScoped.class)
+        .inject(this)
+        .build();
+
+    @Test
+    public void someTest() {
+        // ...
+    }
+}
+----
+
+That's very convenient, but it might cause some slowness for larger classpaths.
+So alternatively, all bean types to be used during the test can be passed explicitly:
+
+[source,java]
+----
+@Rule
+public WeldInitiator weld = WeldInitiator.from(
+        GreetingService.class,
+        TransactionalGreetingService.class,
+        EntityManagerProducer.class,
+        EntityManagerFactoryProducer.class,
+        TransactionExtension.class,
+        // ...
+    )
+    .activate(RequestScoped.class)
+    .inject(this)
+    .build();
+----
+
+This avoids the classpath scanning but comes at the cost of increased efforts for writing and maintaining the test.
+My recommendation would be to go for the classpath scanning approach and only switch over to explicitly listing all classes if the scanning actually isn't feasible.
```

Comment:
For your case it might be easier to just register a package with Weld (assuming all those classes are in the same test package, or in two).
You could use either `WeldInitiator.ofTestPackage()` to automatically add only classes from test package, or you could leverage what Weld SE offers since you are already using `new Weld()` anyway.
Something like - `WeldInitiator.from(new Weld().addPackages(GreetingService.class.getPackage()))`

Reply:
Reworded.